File: src/main/java/com/mpe85/grampa/rule/impl/CharPredicateRule.java
package com.mpe85.grampa.rule.impl;

import static com.google.common.base.Preconditions.checkNotNull;

import java.util.Objects;
import java.util.function.Predicate;

import com.google.common.base.MoreObjects.ToStringHelper;
import com.mpe85.grampa.rule.RuleContext;

/**
 * A character predicate rule implementation.
 * 
 * @author mpe85
 *
 * @param <T>
 *            the type of the stack elements
 */
public class CharPredicateRule<T> extends AbstractRule<T> {
	
	/**
	 * C'tor. Create a character predicate rules that exactly matches a specific character.
	 * 
	 * @param character
	 *            a character
	 */
	public CharPredicateRule(final char character) {
		this(c -> c == character);
	}
	
	/**
	 * C'tor. Create a character predicate rule.
	 * 
	 * @param predicate
	 *            that is tested by the rule.
	 */
	public CharPredicateRule(final Predicate<Character> predicate) {
		this.predicate = checkNotNull(predicate, "A 'predicate' must not be null.");
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		return !context.isAtEndOfInput()
				&& predicate.test(context.getCurrentChar())
				&& context.advanceIndex(1);
	}
	
	@Override
	public int hashCode() {
		return Objects.hash(super.hashCode(), predicate);
	}
	
	@Override
	public boolean equals(final Object obj) {
		if (obj != null && getClass() == obj.getClass()) {
			final CharPredicateRule<?> other = (CharPredicateRule<?>) obj;
			return super.equals(other)
					&& Objects.equals(predicate, other.predicate);
		}
		return false;
	}
	
	@Override
	protected ToStringHelper toStringHelper() {
		return super.toStringHelper()
				.add("predicate", predicate);
	}
	
	
	private final Predicate<Character> predicate;
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/impl/CharPredicateRule.kt
package com.mpe85.grampa.rule.impl

import com.mpe85.grampa.context.ParserContext
import com.mpe85.grampa.rule.Rule
import com.mpe85.grampa.util.checkEquality
import com.mpe85.grampa.util.stringify
import java.lang.Character.toUpperCase
import java.util.Objects.hash

/**
 * A character predicate rule implementation.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @property[predicate] A predicate that is tested by the rule
 */
public open class CharPredicateRule<T>(private val predicate: (Char) -> Boolean) : AbstractRule<T>() {

    /**
     * Construct a character predicate rule that exactly matches a specific character.
     *
     * @param[character] A character
     */
    public constructor(character: Char) : this({ it == character })

    override fun match(context: ParserContext<T>): Boolean = !context.atEndOfInput
            && predicate(context.currentChar)
            && context.advanceIndex(1)

    override fun hashCode(): Int = hash(super.hashCode(), predicate)
    override fun equals(other: Any?): Boolean = checkEquality(other, { super.equals(other) }, { it.predicate })
    override fun toString(): String = stringify("predicate" to predicate)
}

/**
 * An ignore-case character rule implementation.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[character] A character
 */
public class IgnoreCaseCharRule<T>(character: Char) :
    CharPredicateRule<T>({ toUpperCase(it) == toUpperCase(character) })

/**
 * Create a rule from this character.
 *
 * @return A [CharPredicateRule]
 */
public fun <T> Char.toRule(): Rule<T> = CharPredicateRule(this)

/**
 * Create an ignore-case rule from this character.
 *
 * @return A [CharPredicateRule]
 */
public fun <T> Char.toIgnoreCaseRule(): Rule<T> = IgnoreCaseCharRule(this)
